datalad/support/gitrepo.py

```diff
@@ -4076,6 +4056,102 @@ def _save_add(self, files, git_opts=None):
+            v = v.replace('\\', '\\\\')
+            if section:
+                # must not have additional unquoted quotes
+                v = v.replace('"', '\\"')
+            elif v and (v[0] in white or v[-1] in white):
+                # quoting the value due to leading/trailing whitespace
+                # prevent inner unquotes quotes
+                v = v.replace('"', '\\"')
+                v = '"{}"'.format(v)
+            return v
+
+        def write_submodule_config(fobj, name, props):
+            fmt = '[submodule {_q_}{_name_}{_q_}]\n'
+            for p in props:
+                fmt += '\t{p} = {{{p}}}\n'.format(p=p)
+            fobj.write(
+                fmt.format(
+                    _q_='' if name.startswith('"') else '"',
+                    _name_=quote_config(name, section=True),
+                    **{k: quote_config(v) for k, v in props.items()}))
+
+        # first gather info from all datasets in read-only fashion, and then
+        # update index, .gitmodules and .git/config at once
+        info = []
+        for path in paths:
+            rpath = str(path.relative_to(self.pathobj).as_posix())
+            subm = GitRepo(path, create=False, init=False)
+            subm_commit = subm.get_hexsha()
+            if not subm_commit:
+                yield get_status_dict(
+                    action='add_submodule',
+                    ds=self,
+                    path=path,
+                    status='error',
+                    message=('cannot add subdataset %s with no commits', subm),
+                    logger=lgr)
+                continue
+            # make an attempt to configure a submodule source URL based on the
+            # discovered remote configuration
+            remote, branch = subm.get_tracking_branch()
+            url = subm.get_remote_url(remote) if remote else None
+            if url is None:
+                url = './{}'.format(rpath)
+            subm_id = subm.config.get('datalad.dataset.id', None)
+            info.append(
+                dict(path=path, rpath=rpath, commit=subm_commit, id=subm_id,
+                     url=url))
+
+        # bypass any convenience or safe-manipulator for speed reasons
+        # use case: saving many new subdatasets in a single run
+        with (self.pathobj / '.gitmodules').open('a') as gmf, \
+             (self.pathobj / '.git' / 'config').open('a') as gcf:
+            for i in info:
+                self.call_git([
+                    'update-index', '--add', '--replace', '--cacheinfo', '160000',
+                    i['commit'], i['rpath']
+                ])
+                gmprops = dict(path=i['rpath'], url=i['url'])
+                if i['id']:
+                    gmprops['datalad-id'] = i['id']
+                write_submodule_config(
+                    gmf, i['rpath'], gmprops)
+                write_submodule_config(
+                    gcf, i['rpath'], dict(active='true', url=i['url']))
```

Comment:
With this switch to writing the `.git/config` entries on our own, there's a change in how the `url` is represented: before it was an absolute path and now it is relative.  I think that should be fine and for our purposes probably preferable.

Reply:
True. I had the same feeling of "should be better". Related: https://github.com/datalad/datalad/pull/4025#issue-362968783

Reply:
In any case the issue from #3538 remains here, as something earlier makes the path absolute:

```
% dl create raw
% dl create bids
% cd bids
bids % datalad install -d . -s ../raw sourcedata --reckless
bids % cat .gitmodules
[submodule "sourcedata"]
        path = sourcedata
        url = /tmp/d1/bids/../raw
        datalad-id = ca96f420-2f90-4f9c-8c46-0e6213dcfd99
```